File: cookbook/python-supervisor/nodes.py
from brainyflow import Node, Memory # Import Memory
from utils import call_llm, search_web
import yaml
import random

class DecideAction(Node):
    async def prep(self, memory: Memory): # Use memory and add type hint
        """Prepare the context and question for the decision-making process."""
        # Get the current context (default to "No previous search" if none exists)
        context = memory.context if hasattr(memory, 'context') else "No previous search" # Use property access
        # Get the question from memory
        question = memory.question if hasattr(memory, 'question') else "" # Use property access
        # Return both for the exec step
        return question, context

    async def exec(self, inputs):
        """Call the LLM to decide whether to search or answer."""
        question, context = inputs

        print(f"🤔 Agent deciding what to do next...")

        # Create a prompt to help the LLM decide what to do next
        prompt = f"""
### CONTEXT
You are a research assistant that can search the web.
Question: {question}
Previous Research: {context}

### ACTION SPACE
[1] search
  Description: Look up more information on the web
  Parameters:
    - query (str): What to search for

[2] answer
  Description: Answer the question with current knowledge
  Parameters:
    - answer (str): Final answer to the question

## NEXT ACTION
Decide the next action based on the context and available actions.
Return your response in this format:

```yaml
thinking: |
    <your step-by-step reasoning process>
action: search OR answer
reason: <why you chose this action>
search_query: <specific search query if action is search>
```"""

        # Call the LLM to make a decision
        response = call_llm(prompt)
        
        # Parse the response to get the decision
        # Handle potential errors if the response doesn't contain the expected yaml block
        try:
            yaml_str = response.split("```yaml")[1].split("```")[0].strip()
            decision = yaml.safe_load(yaml_str)
        except (IndexError, yaml.YAMLError) as e:
            print(f"Error parsing LLM response: {e}")
            # Return a default or error decision if parsing fails
            return {"action": "error", "reason": f"Failed to parse LLM response: {response[:100]}..."}

        return decision

    async def post(self, memory: Memory, prep_res, exec_res):
        """Save the decision and determine the next step in the flow."""
        # If LLM decided to search, save the search query
        if exec_res.get("action") == "search":
            memory.search_query = exec_res.get("search_query")
            print(f"🔍 Agent decided to search for: {memory.search_query}")
        elif exec_res.get("action") == "answer":
             print(f"💡 Agent decided to answer the question")
        else:
             print(f"⚠️ Agent returned an unexpected action: {exec_res.get('action')}")
             # Handle unexpected actions, maybe trigger an error or fallback
             self.trigger("error")
             return

        # Trigger the action returned by the LLM
        self.trigger(exec_res.get("action"))

class SearchWeb(Node):
    async def prep(self, memory: Memory):
        """Get the search query from memory."""
        return memory.search_query if hasattr(memory, 'search_query') else ""

    async def exec(self, search_query):
        """Search the web for the given query."""
        # Call the search utility function
        print(f"🌐 Searching the web for: {search_query}")
        results = await search_web(search_query)
        return results

    async def post(self, memory: Memory, prep_res, exec_res):
        """Save the search results and go back to the decision node."""
        # Add the search results to the context in memory
        previous = memory.context if hasattr(memory, 'context') else ""
        memory.context = previous + "\n\nSEARCH: " + (memory.search_query if hasattr(memory, 'search_query') else "") + "\nRESULTS: " + exec_res

        print(f"📚 Found information, analyzing results...")

        # Always go back to the decision node after searching
        self.trigger("decide")

class UnreliableAnswerNode(Node):
    async def prep(self, memory: Memory):
        """Get the question and context for answering."""
        return memory.question if hasattr(memory, 'question') else "", memory.context if hasattr(memory, 'context') else ""

    async def exec(self, inputs):
        """Call the LLM to generate a final answer with 50% chance of returning a dummy answer."""
        question, context = inputs

        # 50% chance to return a dummy answer
        if random.random() < 0.5:
            print(f"🤪 Generating unreliable dummy answer...")
            return "Sorry, I'm on a coffee break right now. All information I provide is completely made up anyway. The answer to your question is 42, or maybe purple unicorns. Who knows? Certainly not me!"

        print(f"✍️ Crafting final answer...")

        # Create a prompt for the LLM to answer the question
        prompt = f"""
### CONTEXT
Based on the following information, answer the question.
Question: {question}
Research: {context}

## YOUR ANSWER:
Provide a comprehensive answer using the research results.
"""
        # Call the LLM to generate an answer
        answer = call_llm(prompt)
        return answer

    async def post(self, memory: Memory, prep_res, exec_res):
        """Save the final answer and complete the flow."""
        # Save the answer in memory
        memory.answer = exec_res

        print(f"✅ Answer generated successfully")
        self.trigger("default")

class SupervisorNode(Node):
    async def prep(self, memory: Memory):
        """Get the current answer for evaluation."""
        return memory.answer if hasattr(memory, 'answer') else None

    async def exec(self, answer):
        """Check if the answer is valid or nonsensical."""
        if answer is None:
             print(f"    🔍 Supervisor received no answer to check.")
             return {"valid": False, "reason": "No answer provided"}

        print(f"    🔍 Supervisor checking answer quality...")

        # Check for obvious markers of the nonsense answers
        nonsense_markers = [
            "coffee break",
            "purple unicorns",
            "made up",
            "42",
            "Who knows?"
        ]

        # Check if the answer contains any nonsense markers
        is_nonsense = any(marker in answer for marker in nonsense_markers)

        if is_nonsense:
            return {"valid": False, "reason": "Answer appears to be nonsensical or unhelpful"}
        else:
            return {"valid": True, "reason": "Answer appears to be legitimate"}

    async def post(self, memory: Memory, prep_res, exec_res):
        """Decide whether to accept the answer or restart the process."""
        if exec_res.get("valid", False):
            print(f"    ✅ Supervisor approved answer: {exec_res.get('reason', 'N/A')}")
            self.trigger("approved")
        else:
            print(f"    ❌ Supervisor rejected answer: {exec_res.get('reason', 'N/A')}")
            # Clean up the bad answer
            memory.answer = None
            # Add a note about the rejected answer
            context = memory.context if hasattr(memory, 'context') else ""
            memory.context = context + "\n\nNOTE: Previous answer attempt was rejected by supervisor."
            self.trigger("retry")

File: cookbook/typescript-agent/nodes.ts
import { Memory, Node } from 'caskada'
import { parse } from 'yaml'
import { callLLM, webSearch } from './utils'

export type SearchAgentGlobalStore = {
  question: string
  context?: string
  searchQuery?: string
  answer?: string
}

interface PrepResult {
  question: string
  context?: string
}

// Define allowed actions for DecideNode
type DecideNodeActions = ['search', 'answer', 'error']

interface LLMDecision {
  thinking: string
  action: 'search' | 'answer'
  reason: string
  answer?: string
  searchQuery?: string
}

interface SearchResult {
  title: string
  href: string
  body: string
}

export class DecideNode extends Node<SearchAgentGlobalStore, PrepResult, LLMDecision, DecideNodeActions> {
  async prep(memory: Memory<SearchAgentGlobalStore>) {
    const context = memory.context || 'No previous search.'
    const question = memory.question
    return { context, question }
  }

  async exec(input: PrepResult) {
    const { context, question } = input
    const now = new Date()
    console.log('Deciding whether to search or answer the question...')

    const prompt = `
        ### Context
        Current date is ${now.toISOString()}
        You are helpful assistance that can searching the web to gather real time data.
        Question: ${question}
        Context: ${context}

        ### Action Space
        [1] search
            description: look up for more information on internet
            parameter:
                - query (str): what to search for
        [2] answer
            description: answer the question based on the context
            parameter:
                - answer (str): the answer to the question
        
        ### Next Action
        Decide the next action based on the context and available actions.
        Return your response in this format:

        \`\`\`yaml
        thinking: |
            <your step-by-step reasoning process>
        action: search OR answer
        reason: <why you chose this action>
        answer: <if action is answer>
        searchQuery: <specific search query if action is search>
        \`\`\`

        IMPORTANT: Make sure to:
        1. Use proper indentation (4 spaces) for all multi-line fields
        2. Use the | character for multi-line text fields
        3. Keep single-line fields without the | character
        4. thinking about search query, make sure that you understand the question and use query that appropriate for search engine, not just copying the question
        `
    const response = await callLLM([{ role: 'user', content: prompt }])
    if (!response?.includes('```yaml')) {
      throw new Error('LLM response missing YAML block')
    }

    const yamlStr = response.split('```yaml')[1].split('```')[0].trim()
    return parse(yamlStr) as LLMDecision
  }

  async post(memory: Memory<SearchAgentGlobalStore>, prepRes: PrepResult, execRes: LLMDecision) {
    if (!prepRes) {
      console.log('No context or question provided.')
      this.trigger('error')
      return
    }

    if (!execRes) {
      console.log('No decision made.')
      this.trigger('error')
      return
    }

    if (execRes.action === 'search') {
      memory.searchQuery = execRes.searchQuery
      console.log(`Searching for: ${execRes.searchQuery}`)
      console.log(`Reason: ${execRes.reason}`)
    } else {
      memory.answer = execRes.answer // Persist final answer
      console.log(`Answering: ${execRes.answer}`)
      console.log(`Reason: ${execRes.reason}`)
    }

    this.trigger(execRes.action)
  }
}

// Define allowed actions for SearchNode
type SearchNodeActions = ['decide', 'error']

export class SearchNode extends Node<SearchAgentGlobalStore, string, SearchResult[], SearchNodeActions> {
  async prep(memory: Memory<SearchAgentGlobalStore>) {
    return memory.searchQuery
  }

  async exec(searchQuery: string) {
    console.log(`Calling web search tool.`)
    return await webSearch(searchQuery)
  }

  async post(memory: Memory<SearchAgentGlobalStore>, prepRes: string, execRes: SearchResult[]) {
    if (!prepRes) {
      console.log('No search query provided.')
      this.trigger('error')
      return
    }

    if (!execRes) {
      console.log('No search results found.')
      this.trigger('error')
      return
    }

    const previous = memory.context || ''
    memory.context = previous + '\n\nSearch: ' + memory.searchQuery + '\nResult :' + JSON.stringify(execRes)
    this.trigger('decide')
  }
}

// Define allowed actions for AnswerNode
type AnswerNodeActions = ['done', 'error']

export class AnswerNode extends Node<SearchAgentGlobalStore, PrepResult, string | null, AnswerNodeActions> {
  async prep(memory: Memory<SearchAgentGlobalStore>) {
    const context = memory.context || 'No previous context.'
    const question = memory.question
    return { question, context }
  }

  async exec(input: PrepResult) {
    const { question, context } = input
    console.log('Answering the question...')
    const prompt = `
        ### Context
        Based on the following information, answer the question.
        Question: ${question}
        Research: ${context}
        
        ## Your Answer:
        Provide a comprehensive answer using the research results.
        `
    const response = await callLLM([{ role: 'user', content: prompt }])
    return response
  }

  async post(memory: Memory<SearchAgentGlobalStore>, prepRes: PrepResult, execRes: string | null) {
    if (!prepRes) {
      console.log('No answer provided.')
      this.trigger('error')
      return
    }

    if (!execRes) {
      console.log('No answer generated.')
      this.trigger('error')
      return
    }

    memory.answer = execRes
    console.log(`Final Answer: ${execRes}`)
    this.trigger('done')
  }
}
